sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java

```diff
@@ -101,7 +101,7 @@ public Duration getAllowedTimestampSkew() {
   public PCollection<T> apply(PCollection<T> input) {
     return input
         .apply("AddTimestamps", ParDo.of(new AddTimestampsDoFn<T>(fn, allowedTimestampSkew)))
-        .setTypeDescriptorInternal(input.getTypeDescriptor());
+        .setCoder(input.getCoder());
```

Comment:
If a coder is explicitly set on the input PCollection (or really any upstream PCollection) we should respect that rather than performing coder inference again.